### PR TITLE
feat: whitelist Govsg templates on per user basis

### PR DESCRIPTION
**Pull request by tan-zx**
Todo:
- [x] Migration script to grant permission to existing users 
- [x] Test on staging

Deployment notes
- [x] Run db migration
- [x] Grant permission to new table (run those Postgres commands)
- [x] Run migration script to ensure current users retain existing permissions
- [x] Remove script from this PR before merging into db
- [ ] Deploy to prod

### CODE REVIEW

File: backend/src/database/migrations/20230710111307-create-govsg-template-access-table.js [hunk 1]
```diff
@@ -0,0 +1,46 @@
+'use strict';
+
+module.exports = {
+  up: async (queryInterface, Sequelize) => {
+    queryInterface.createTable('govsg_templates_access', {
```
Comment (stanleynguyen) [inline backend/src/database/migrations/20230710111307-create-govsg-template-access-table.js:5]: missing await

File: backend/src/database/migrations/20230710111307-create-govsg-template-access-table.js [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+      created_at: {
+        type: Sequelize.DataTypes.DATE,
+        allowNull: false,
+      },
+      updated_at: {
+        type: Sequelize.DataTypes.DATE,
+        allowNull: false,
+      },
+    })
+    queryInterface.addIndex('govsg_templates_access', ['template_id', 'user_id'], {
```
Comment (stanleynguyen) [inline backend/src/database/migrations/20230710111307-create-govsg-template-access-table.js:37]: missing await

File: backend/src/govsg/middlewares/govsg-template.middleware.ts [hunk 3]
```diff
@@ -21,19 +23,47 @@ import { NextFunction, Request, Response } from 'express'
...
+      `User ${userId} does not have access to template ${templateId}`
+    )
+  }
+}
+
 export async function pickTemplateForCampaign(
   req: Request,
   res: Response
 ): Promise<Response> {
+  if (!req.session?.user?.id) {
```
Comment (stanleynguyen) [inline backend/src/govsg/middlewares/govsg-template.middleware.ts:60]: no need to add this, we already have an authentication middle ware installed in the route iinm

Comment (tan-zx) [reply backend/src/govsg/middlewares/govsg-template.middleware.ts:60]: yeah it's not necessary but I was just following [the code here hmm](https://github.com/opengovsg/postmangovsg/blob/dc4785a7c1f53972ad181537a57c580df2d603d8/backend/src/core/middlewares/experiment.middleware.ts)

File: backend/src/database/migrations/20230710111307-create-govsg-template-access-table.js [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+      unique: true,
+      name: 'govsg_templates_access_template_id_user_id_unique'
+    })
+  },
+
+  down: async (queryInterface, _) => {
+    queryInterface.dropTable('govsg_templates_access')
```
[the remark this answers is not in the document]
Comment (tan-zx) [reply backend/src/database/migrations/20230710111307-create-govsg-template-access-table.js]: whenever I write migrations, I really hate how there isn't TS support...

**Comment by tan-zx** [comment]
not sure if these should be `string` instead of `number`

image
image

**Comment by tan-zx** [comment]
use composite primary key instead for simplicity 
image

File: backend/src/govsg/middlewares/govsg-template.middleware.ts (hunk 3 above)
Comment (stanleynguyen) [reply backend/src/govsg/middlewares/govsg-template.middleware.ts:60]: oh lmao yea let's remove both

File: backend/src/database/migrations/20230710111307-create-govsg-template-access-table.js [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+      },
+      created_at: {
+        type: Sequelize.DataTypes.DATE,
+        allowNull: false,
+      },
+      updated_at: {
+        type: Sequelize.DataTypes.DATE,
+        allowNull: false,
+      },
+    })
```
Comment (stanleynguyen) [inline backend/src/database/migrations/20230710111307-create-govsg-template-access-table.js:34]: I think should add an index `unique(template_id, user_id)` here so can search faster, and also to enforce uniqueness

Comment (tan-zx) [reply backend/src/database/migrations/20230710111307-create-govsg-template-access-table.js:34]: if it's alr a composite primary key, then there's alr a uniqueness constraint right 🤔️

File: backend/src/govsg/middlewares/govsg-template.middleware.ts (hunk 3 above)
Comment (tan-zx) [reply backend/src/govsg/middlewares/govsg-template.middleware.ts:60]: OK done

**Review by stanleynguyen — APPROVED** [review]

**Pull request merged**